cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -478,40 +480,84 @@ public Map<ApplicationId, ApplicationMeta> getApplicationsForAppIds(Collection<A
    *
-   * @param programIds the set of program ids to filter
+   * @param programRefs the set of program ids to filter
    * @return the set of program ids that exist
    * @throws IOException if failed to read metadata
    */
-  public Set<ProgramId> filterProgramsExistence(Collection<ProgramId> programIds) throws IOException {
-    Set<ApplicationId> appIds = programIds.stream().map(ProgramId::getParent).collect(Collectors.toSet());
-    List<List<Field<?>>> multiKeys = new ArrayList<>();
-    for (ApplicationId appId: appIds) {
-      multiKeys.add(getApplicationPrimaryKeys(appId));
+  public Set<ProgramId> filterProgramsExistence(Collection<ProgramReference> programRefs) throws IOException {
+    Set<ApplicationReference> appRefs = programRefs.stream()
+      .map(ProgramReference::getParent).collect(Collectors.toSet());
+    List<Range> multiRanges = new ArrayList<>();
+    for (ApplicationReference appRef: appRefs) {
+      // Create Scan ranges to get the "latest" and "-SNAPSHOT" versions
+      multiRanges.add(Range.singleton(getLatestApplicationKeys(appRef)));
+      multiRanges.add(Range.singleton(getApplicationPrimaryKeys(appRef.app(ApplicationId.DEFAULT_VERSION))));
     }
 
-    Set<ProgramId> existingPrograms = new HashSet<>();
-    for (StructuredRow row : getApplicationSpecificationTable().multiRead(multiKeys)) {
-      ApplicationId appId = getApplicationIdFromRow(row);
-      String appMeta = row.getString(StoreDefinition.AppMetadataStore.APPLICATION_DATA_FIELD);
-      if (appMeta == null) {
-        throw new IOException("Missing application metadata for application " + appId);
-      }
-      try (JsonReader reader = new JsonReader(new StringReader(appMeta))) {
-        reader.beginObject();
-        while (reader.peek() != JsonToken.END_OBJECT) {
-          String name = reader.nextName();
-          if (name.equals("spec")) {
-            existingPrograms.addAll(ApplicationSpecificationAdapter.getProgramIds(appId, reader));
-          } else {
-            reader.skipValue();
+    // The latest version programs
+    Map<ApplicationReference, Set<ProgramId>> latestProgramIdsMap = new HashMap<>();
+    // The latest versions
+    Map<ApplicationReference, String> latestAppVersions = new HashMap<>();
+    try (CloseableIterator<StructuredRow> iterator =
+           getApplicationSpecificationTable().multiScan(multiRanges, appRefs.size() * 2)) {
+      while (iterator.hasNext()) {
+        StructuredRow row = iterator.next();
+        ApplicationId appId = getApplicationIdFromRow(row);
+        String appMeta = row.getString(StoreDefinition.AppMetadataStore.APPLICATION_DATA_FIELD);
+        if (appMeta == null) {
+          throw new IOException("Missing application metadata for application " + appId);
+        }
+        
+        ApplicationReference appRef = appId.getAppReference();
+        String isLatest = row.getString(StoreDefinition.AppMetadataStore.LATEST_FIELD);
```

Comment:
I thought isLatest is a boolean column?

Reply:
I think we did this intentionally, because spi doesn't have boolean field type https://github.com/cdapio/cdap/blob/develop/cdap-storage-spi/src/main/java/io/cdap/cdap/spi/data/table/field/Fields.java

Reply:
I see. Can we add a boolean type? Don't see why it shouldn't be there.

Reply:
@albertshau Yeah that we can do: https://cdap.atlassian.net/browse/CDAP-19981, but out of scope for now

Reply:
note that this should be done before 6.8.0, as changing the type of an existing column is not a trivial schema change

Reply:
So I think we kept it in the design doc. What is the cons of using it as a string? If we determine to add the boolean type now, we need to do it right away

Reply:
I am not sure if we are able to accomadate that for 6.8. And as @de-lan mentioned, is there a disadvantage of using it as a String?

Reply:
the disadvantage is in size/performance and correctness. With a boolean, you don't ever have to worry about bad data somehow getting in there, whether by buggy code, manual intervention, etc.